monitoring: add ceph_nvmeof_metdata to ceph-exporter

VallariAg opened the pull request:
This PR also fixes this bug:
If more than 1 gateway is down (i.e. in UNAVAILABLE state), NVMeoF Overview grafana charts for "Total Gateways" and "Ceph Health NVMeoF WARNING" graphs show wrong information. 
Suppose in 4 gateway env, if 2 gw are down then "Total Gateways" grafana graph shows "Total 3, Available 2, Down 1", instead of the correct "Total 4, Available 2, Down 2". 
Also in same env, graph "Ceph Health NVMeoF WARNING" shows 1 count of "NVMEOF_GATEWAY_DOWN", even when multiple gateways are down. 
BZ: https://bugzilla.redhat.com/show_bug.cgi?id=2364869#

Before:
<img width="1483" alt="Screenshot 2025-05-19 at 9 58 07 PM" src="https://github.com/user-attachments/assets/9ccc7040-b7a6-4651-9702-e9e255b9067a" />

After:
<img width="1484" alt="Screenshot 2025-05-19 at 10 00 24 PM" src="https://github.com/user-attachments/assets/84fd2884-e41a-4560-9421-bc6c54f1462a" />



<!--


  - Please give your pull request a title like

      [component]: [short description]

  - Please use this format for each git commit message:

      [component]: [short description]

      [A longer multiline description]

      Fixes: [ticket URL on tracker.ceph.com, create one if necessary]
      Signed-off-by: [Your Name] <[your email]>

    For examples, use "git log".
-->

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [x] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins test classic perf` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-perf-classic/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-perf-pull-requests/config/definitions/ceph-perf-pull-requests.yml)
- `jenkins test crimson perf` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-perf-crimson/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-perf-pull-requests/config/definitions/ceph-perf-pull-requests.yml)
- `jenkins test signed` [Jenkins Job](https://jenkins.ceph.com/job/ceph-pr-commits/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pr-commits/config/definitions/ceph-pr-commits.yml)
- `jenkins test make check` [Jenkins Job](https://jenkins.ceph.com/job/ceph-pull-requests/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pull-requests/config/definitions/ceph-pull-requests.yml)
- `jenkins test make check arm64` [Jenkins Job](https://jenkins.ceph.com/job/ceph-pull-requests-arm64/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pull-requests-arm64/config/definitions/ceph-pull-requests-arm64.yml)
- `jenkins test submodules` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-pr-submodules/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pr-submodules/config/definitions/ceph-pr-commits.yml)
- `jenkins test dashboard` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-dashboard-pull-requests/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-dashboard-pull-requests/config/definitions/ceph-dashboard-pull-requests.yml)
- `jenkins test dashboard cephadm` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-dashboard-cephadm-e2e/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-dashboard-cephadm-e2e/config/definitions/ceph-dashboard-cephadm-e2e.yml)
- `jenkins test api` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-api/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pr-api/config/definitions/ceph-pr-api.yml)
- `jenkins test docs` [ReadTheDocs](https://readthedocs.org/projects/ceph/) | [Github Workflow Definition](https://github.com/ceph/ceph/blob/main/.readthedocs.yml)
- `jenkins test ceph-volume all` [Jenkins Jobs](https://jenkins.ceph.com/view/ceph-volume%20PR/) | [Jenkins Jobs Definition](https://github.com/ceph/ceph-build/blob/main/ceph-volume-cephadm-prs/config/definitions/ceph-volume-pr.yml)
- `jenkins test windows` [Jenkins Job](https://jenkins.ceph.com/job/ceph-windows-pull-requests/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-windows-pull-requests/config/definitions/ceph-windows-pull-requests.yml)
- `jenkins test rook e2e` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-orchestrator-rook-e2e/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-rook-e2e/config/definitions/ceph-orchestrator-rook-e2e.yml)
</details>

cloudbehl commented:
@VallariAg you will need to create upstream tracker for this

VallariAg commented:
Sure, I created one here: https://tracker.ceph.com/issues/71384 
Please see if it looks okay. Does this PR also needs to be backported to tentacle branch after merge?

cloudbehl commented:
> Sure, I created one here: https://tracker.ceph.com/issues/71384
You need to add this to you PR description as well.

> Please see if it looks okay. Does this PR also needs to be backported to tentacle branch after merge?
Yes it would be good to get this backported as well, otherwise user will hit the issue and we will see the issues being created

cloudbehl commented on src/pybind/mgr/prometheus/module.py:
```diff
@@ -740,6 +742,13 @@ def _setup_static_metrics(self) -> Dict[str, Metric]:
             RBD_MIRROR_METADATA
         )
 
+        metrics['nvmeof_metadata'] = Metric(
+            'untyped',
+            'nvmeof_metadata',
+            'NVMeoF Metadata',
+            NVMEOF_METADATA
+        )
+
```
@VallariAg like NVME can have different kind of metadata. 

NVME gateway group metadata 
NVME gateway metadata.

I think what you are trying to expose if daemon status of gateway, if I'm not wrong. So lets make this status metric rather than metadata, is what I'm trying to say. i.e NVMEOF_GATEWAY_STATUS

In this case the status would be output of the metric and not the label.

afreen23 commented:
jenkins test dashboard cephadm

VallariAg commented:
@cloudbehl I have changed the metric so the value is the status, instead of "status" label. 

Updated the timeseries graph to also show which gateways are down (along with its bad state):

<img width="1167" height="359" alt="Screenshot 2025-08-11 at 5 25 49 PM" src="https://github.com/user-attachments/assets/dd931773-e09c-4a60-b4bd-cbd2f897c6ad" />

cloudbehl commented:
@VallariAg I have written this PR, I think we can take this forward and the nvme daemons can be filtered from this. What do you thing?

https://github.com/ceph/ceph/pull/64912

cloudbehl commented on monitoring/ceph-mixin/dashboards/ceph-nvmeof.libsonnet:
```diff
@@ -119,191 +119,164 @@ local g = import 'grafonnet/grafana.libsonnet';
-        datasource='$datasource',
+      .addTarget(
+        $.addTargetSchema(
+          expr='group by (daemon_name) (ceph_cephadm_daemon_status{service_type="nvmeof"} == -1)',
```
I have not documented -1, -2 in my PR, are you sure, this is what they are reporting?

cloudbehl commented on monitoring/ceph-mixin/dashboards/ceph-nvmeof.libsonnet:
```diff
@@ -119,191 +119,164 @@ local g = import 'grafonnet/grafana.libsonnet';
-        legendFormat='NVMEOF_GATEWAY_DOWN',
-        range=true,
+      $.timeSeriesPanel( 
+        title='Gateway WARNING Health States',
```
We can include the healthy one's as well. Maybe call it `Gateway health status`?

VallariAg replied:
The graph shows total number of gateways in bad state (and shows daemons names and the state on hover). That's why I'm not sure how we can include the healthy ones here. If we show all states, then unhealthy states would not be highlighted. 

<img width="1193" height="883" alt="Screenshot 2025-10-08 at 1 12 28 PM" src="https://github.com/user-attachments/assets/1f835b1d-31be-4354-b3ee-b0e2bdc0b402" />

cloudbehl replied:
The thing is even though description of pane says Warning health states -> The stopped or error are danger/error state.  So the Panel name looks confusing.

I think we don't need graph for representation the state. Its a point in time data, so single stat should be good. I find it confusing to read personally.

VallariAg commented on monitoring/ceph-mixin/dashboards/ceph-nvmeof.libsonnet:
```diff
@@ -119,191 +119,164 @@ local g = import 'grafonnet/grafana.libsonnet';
-        datasource='$datasource',
+      .addTarget(
+        $.addTargetSchema(
+          expr='count(ceph_cephadm_daemon_status{service_type="nvmeof"}==1 or ceph_cephadm_daemon_status{service_type="nvmeof"}==2)',
```
So I tested for status=2 (starting) and status=1 (running):
<img width="1728" height="201" alt="Screenshot 2025-10-08 at 12 50 07 PM" src="https://github.com/user-attachments/assets/3cb3c64d-8506-4693-8556-1167b165eff9" />

<img width="1470" height="123" alt="Screenshot 2025-10-08 at 12 50 21 PM" src="https://github.com/user-attachments/assets/e662f7f4-d9ac-4647-ae3f-13c968673812" />

And status=0 (stopped)
<img width="1727" height="204" alt="Screenshot 2025-10-08 at 1 05 08 PM" src="https://github.com/user-attachments/assets/39e07838-099d-4aad-8da7-a301f230e1b8" />

<img width="1468" height="113" alt="Screenshot 2025-10-08 at 1 05 26 PM" src="https://github.com/user-attachments/assets/3b413bd4-5a9c-4fbd-854a-3cc9c6afd09b" />


So I think above mapping can be trusted? What do you think?

VallariAg commented:
@cloudbehl Ohh I see. Will look into it, thank you!

VallariAg commented:
@cloudbehl, please help me understand the failures in the dashboard CI checks.
